File: src/sandbox/backends/e2b.ts
import { performance } from "node:perf_hooks";

import type {
  SandboxBackend,
  SandboxBackendCapabilities,
  SandboxExecutionRequest,
  SandboxExecutionResult,
  SandboxFileEntry,
  SandboxSession,
  SandboxSnapshot,
} from "../types.js";

type E2BExecutorResult = {
  exitCode?: number;
  stdout?: string;
  stderr?: string;
};

type E2BCommandResult = {
  exitCode?: number;
  stdout?: string;
  stderr?: string;
};

type E2BSandboxInstance = {
  sandboxId: string;
  commands: {
    run: (cmd: string, opts?: Record<string, unknown>) => Promise<E2BCommandResult>;
  };
  files?: {
    write: (path: string, content: string) => Promise<void>;
    read: (path: string) => Promise<string>;
  };
  kill: (opts?: Record<string, unknown>) => Promise<void>;
  pause?: (opts?: Record<string, unknown>) => Promise<boolean>;
  createSnapshot?: (opts?: Record<string, unknown>) => Promise<{ snapshotId: string }>;
};

type E2BSandboxStatic = {
  create: (...args: unknown[]) => Promise<E2BSandboxInstance>;
  connect: (sandboxId: string, opts?: Record<string, unknown>) => Promise<E2BSandboxInstance>;
  kill?: (sandboxId: string, opts?: Record<string, unknown>) => Promise<boolean>;
};

type E2BModule = {
  Sandbox: E2BSandboxStatic;
};

export interface E2BSandboxBackendOptions {
  apiKey?: string;
  available?: boolean;
  executor?: (
    request: SandboxExecutionRequest,
    session?: SandboxSession,
  ) => Promise<E2BExecutorResult>;
  fileWriter?: (
    files: SandboxFileEntry[],
    session?: SandboxSession,
  ) => Promise<void>;
  fileReader?: (
    path: string,
    session?: SandboxSession,
  ) => Promise<string>;
  template?: string;
  timeoutMs?: number;
}

function quoteShell(value: string): string {
  return `'${value.replaceAll("'", `'\\''`)}'`;
}

function buildShellCommand(request: SandboxExecutionRequest): string {
  const segments: string[] = [];
  if (request.cwd) {
    segments.push(`cd ${quoteShell(request.cwd)}`);
  }
  const argv = [request.command, ...(request.args ?? [])].map(quoteShell).join(" ");
  segments.push(argv);
  return segments.join(" && ");
}

function normalizeEnv(env?: Record<string, string>): Record<string, string> {
  if (!env) {
    return {};
  }

  return Object.fromEntries(
    Object.entries(env).filter((entry): entry is [string, string] => {
      const [key, value] = entry;
      return key.length > 0 && typeof value === "string";
    }),
  );
}

export class E2BSandboxBackend implements SandboxBackend {
  public readonly kind = "e2b" as const;
  private sdkModulePromise: Promise<E2BModule> | null = null;

  public constructor(private readonly options: E2BSandboxBackendOptions = {}) {}

  public capabilities(): SandboxBackendCapabilities {
    return {
      kind: this.kind,
      supportsPersistence: true,
      supportsSnapshots: true,
      supportsBrowser: true,
      supportsNetwork: true,
      supportsPackageInstall: true,
      supportedCapabilities: [
        "command-exec",
        "filesystem",
        "network",
        "browser",
        "package-install",
        "snapshot",
        "resume",
      ],
    };
  }

  public async isAvailable(): Promise<boolean> {
    if (typeof this.options.available === "boolean") {
      return this.options.available;
    }

    if (!this.resolveApiKey()) {
      return false;
    }

    try {
      await this.loadSdk();
      return true;
    } catch {
      return false;
    }
  }

  public async createSession(metadata?: Record<string, string>): Promise<SandboxSession> {
    const sandbox = await this.createSandbox(metadata);
    return {
      id: sandbox.sandboxId,
      backend: this.kind,
      metadata,
    };
  }

  public async destroySession(sessionId: string): Promise<void> {
    const sandbox = await this.connectSandbox(sessionId);
    await sandbox.kill();
  }

  public async writeFiles(
    files: SandboxFileEntry[],
    session?: SandboxSession,
  ): Promise<void> {
    if (this.options.fileWriter) {
      await this.options.fileWriter(files, session);
      return;
    }

    await this.withSandbox(session, async (sandbox) => {
      if (!sandbox.files?.write) {
        throw new Error("Installed E2B SDK does not expose files.write()");
      }

      for (const file of files) {
        await sandbox.files.write(file.path, file.content);
      }
    });
  }

  public async readFile(path: string, session?: SandboxSession): Promise<string> {
    if (this.options.fileReader) {
      return await this.options.fileReader(path, session);
    }

    return await this.withSandbox(session, async (sandbox) => {
      if (!sandbox.files?.read) {
        throw new Error("Installed E2B SDK does not expose files.read()");
      }

      return await sandbox.files.read(path);
    });
  }

  public async execute(
    request: SandboxExecutionRequest,
    session?: SandboxSession,
  ): Promise<SandboxExecutionResult> {
    const startedAt = performance.now();

    try {
      const result = this.options.executor
        ? await this.options.executor(request, session)
        : await this.executeWithE2B(request, session);

      return {
        backend: this.kind,
        exitCode: result.exitCode ?? 0,
        stdout: result.stdout ?? "",
        stderr: result.stderr ?? "",
        durationMs: performance.now() - startedAt,
      };
    } catch (error) {
      return {
        backend: this.kind,
        exitCode: 1,
        stdout: "",
        stderr: error instanceof Error ? error.message : String(error),
        durationMs: performance.now() - startedAt,
      };
    }
  }

  public async snapshot(session: SandboxSession): Promise<SandboxSnapshot> {
    const sandbox = await this.connectSandbox(session.id);
    if (!sandbox.createSnapshot) {
      throw new Error("Installed E2B SDK does not expose createSnapshot()");
    }

    const result = await sandbox.createSnapshot();
    return {
      id: result.snapshotId,
      createdAt: new Date().toISOString(),
      backend: this.kind,
    };
  }

  public async resume(snapshotId: string): Promise<SandboxSession> {
    const sdk = await this.loadSdk();
    const sandbox = this.options.template
      ? await sdk.Sandbox.create(this.options.template, {
          apiKey: this.resolveApiKey(),
          timeoutMs: this.options.timeoutMs,
          metadata: { sourceSnapshotId: snapshotId },
        })
      : await sdk.Sandbox.create(snapshotId, {
          apiKey: this.resolveApiKey(),
          timeoutMs: this.options.timeoutMs,
          metadata: { sourceSnapshotId: snapshotId },
        });

    return {
      id: sandbox.sandboxId,
      backend: this.kind,
      metadata: { sourceSnapshotId: snapshotId },
    };
  }

  private async executeWithE2B(
    request: SandboxExecutionRequest,
    session?: SandboxSession,
  ): Promise<E2BExecutorResult> {
    return await this.withSandbox(session, async (sandbox) => {
      const result = await sandbox.commands.run(buildShellCommand(request), {
        cwd: request.cwd,
        envs: normalizeEnv(request.env),
        timeoutMs: request.timeoutMs ?? this.options.timeoutMs ?? 60_000,
      });
      return {
        exitCode: result.exitCode ?? 0,
        stdout: result.stdout ?? "",
        stderr: result.stderr ?? "",
      };
    });
  }

  private async createSandbox(
    metadata?: Record<string, string>,
  ): Promise<E2BSandboxInstance> {
    const sdk = await this.loadSdk();
    const opts = {
      apiKey: this.resolveApiKey(),
      timeoutMs: this.options.timeoutMs,
      metadata,
    };

    if (this.options.template) {
      return await sdk.Sandbox.create(this.options.template, opts);
    }

    return await sdk.Sandbox.create(opts);
  }

  private async connectSandbox(sessionId: string): Promise<E2BSandboxInstance> {
    const sdk = await this.loadSdk();
    return await sdk.Sandbox.connect(sessionId, {
      apiKey: this.resolveApiKey(),
      timeoutMs: this.options.timeoutMs,
    });
  }

  private async loadSdk(): Promise<E2BModule> {
    if (!this.sdkModulePromise) {
      this.sdkModulePromise = import("e2b")
        .then((module) => module as E2BModule)
        .catch((error) => {
          throw new Error(
            `Failed to load E2B SDK. Install it with 'npm install e2b'. ${error instanceof Error ? error.message : String(error)}`,
          );
        });
    }

    return await this.sdkModulePromise;
  }

  private resolveApiKey(): string | undefined {
    return this.options.apiKey ?? process.env.E2B_API_KEY;
  }

  private async withSandbox<T>(
    session: SandboxSession | undefined,
    action: (sandbox: E2BSandboxInstance) => Promise<T>,
  ): Promise<T> {
    const ephemeral = !session;
    const sandbox = session ? await this.connectSandbox(session.id) : await this.createSandbox();

    try {
      return await action(sandbox);
    } finally {
      if (ephemeral) {
        await sandbox.kill().catch(() => undefined);
      }
    }
  }
}

File: titanx/sandbox/backends/e2b.py
from __future__ import annotations

import time
from dataclasses import dataclass
from datetime import datetime, timezone
from typing import Any, Callable
from uuid import uuid4

from ..types import (
    SandboxBackend,
    SandboxBackendCapabilities,
    SandboxExecutionRequest,
    SandboxExecutionResult,
    SandboxFileEntry,
    SandboxSession,
    SandboxSnapshot,
)


@dataclass
class E2BSandboxBackendOptions:
    api_key: str | None = None
    available: bool = True
    executor: Callable | None = None
    file_writer: Callable | None = None
    file_reader: Callable | None = None


class E2BSandboxBackend(SandboxBackend):
    kind = "e2b"

    def __init__(self, options: E2BSandboxBackendOptions | None = None) -> None:
        self._opts = options or E2BSandboxBackendOptions()
        self._sandbox_instances: dict[str, Any] = {}

    def capabilities(self) -> SandboxBackendCapabilities:
        return SandboxBackendCapabilities(
            kind="e2b",
            supports_persistence=True,
            supports_snapshots=True,
            supports_browser=True,
            supports_network=True,
            supports_package_install=True,
            supported_capabilities=["command-exec", "filesystem", "network", "browser", "snapshot", "resume"],
        )

    async def is_available(self) -> bool:
        if not self._opts.available:
            return False
        try:
            from e2b import Sandbox  # noqa: F401
            return True
        except ImportError:
            return False

    async def execute(
        self,
        request: SandboxExecutionRequest,
        session: SandboxSession | None = None,
    ) -> SandboxExecutionResult:
        start = time.perf_counter()
        try:
            if self._opts.executor:
                res = await self._opts.executor(request, session)
                return SandboxExecutionResult(
                    backend="e2b",
                    exit_code=res.get("exit_code", 0),
                    stdout=res.get("stdout", ""),
                    stderr=res.get("stderr", ""),
                    duration_ms=(time.perf_counter() - start) * 1000,
                )

            from e2b import Sandbox

            cmd_str = " ".join([request.command, *request.args])
            if session and session.id in self._sandbox_instances:
                sbx = self._sandbox_instances[session.id]
            else:
                kwargs: dict[str, Any] = {}
                if self._opts.api_key:
                    kwargs["api_key"] = self._opts.api_key
                sbx = await Sandbox.create(**kwargs)
                if session:
                    self._sandbox_instances[session.id] = sbx

            timeout_s = (request.timeout_ms / 1000) if request.timeout_ms else 60
            result = await sbx.commands.run(cmd_str, timeout=timeout_s, cwd=request.cwd or None)

            return SandboxExecutionResult(
                backend="e2b",
                exit_code=result.exit_code or 0,
                stdout=result.stdout or "",
                stderr=result.stderr or "",
                duration_ms=(time.perf_counter() - start) * 1000,
            )
        except Exception as exc:
            return SandboxExecutionResult(
                backend="e2b", exit_code=1, stdout="", stderr=str(exc),
                duration_ms=(time.perf_counter() - start) * 1000,
            )

    async def create_session(self, metadata: dict[str, str] | None = None) -> SandboxSession:
        from e2b import Sandbox
        kwargs: dict[str, Any] = {}
        if self._opts.api_key:
            kwargs["api_key"] = self._opts.api_key
        sbx = await Sandbox.create(**kwargs)
        session = SandboxSession(id=sbx.sandbox_id, backend="e2b", metadata=metadata or {})
        self._sandbox_instances[session.id] = sbx
        return session

    async def destroy_session(self, session_id: str) -> None:
        sbx = self._sandbox_instances.pop(session_id, None)
        if sbx:
            try:
                await sbx.kill()
            except Exception:
                pass

    async def write_files(
        self, files: list[SandboxFileEntry], session: SandboxSession | None = None
    ) -> None:
        if self._opts.file_writer:
            await self._opts.file_writer(files, session)
            return
        if not session or session.id not in self._sandbox_instances:
            raise ValueError("write_files requires an active E2B session")
        sbx = self._sandbox_instances[session.id]
        for f in files:
            await sbx.files.write(f.path, f.content)

    async def read_file(self, path: str, session: SandboxSession | None = None) -> str:
        if self._opts.file_reader:
            return await self._opts.file_reader(path, session)
        if not session or session.id not in self._sandbox_instances:
            raise ValueError("read_file requires an active E2B session")
        sbx = self._sandbox_instances[session.id]
        return await sbx.files.read(path)

    async def snapshot(self, session: SandboxSession) -> SandboxSnapshot:
        sbx = self._sandbox_instances.get(session.id)
        if not sbx:
            raise ValueError("Cannot snapshot an unknown session")
        if hasattr(sbx, "create_snapshot"):
            result = await sbx.create_snapshot()
            snap_id = result.get("snapshotId", str(uuid4()))
        elif hasattr(sbx, "pause"):
            await sbx.pause()
            snap_id = session.id
        else:
            snap_id = session.id
        return SandboxSnapshot(
            id=snap_id,
            created_at=datetime.now(timezone.utc).isoformat(),
            backend="e2b",
        )

    async def resume(self, snapshot_id: str) -> SandboxSession:
        from e2b import Sandbox
        kwargs: dict[str, Any] = {}
        if self._opts.api_key:
            kwargs["api_key"] = self._opts.api_key
        sbx = await Sandbox.connect(snapshot_id, **kwargs)
        session = SandboxSession(id=sbx.sandbox_id, backend="e2b")
        self._sandbox_instances[session.id] = sbx
        return session
